Add filter list for import and export of extension. Implementation of feature https://github.com/powsybl/powsybl-core/issues/3427

snoopfab opened the pull request:
Describe the motivation
Sometimes we want to import all the extensions except some of them, for that it is easier to propose a blacklist of extensions than having to list all the extensions to import except some of them.
New Parameters :
For Import : iidm.import.xml.filtered-extensions (string list of extensions that will not be imported)
For export iidm.export.xml.filtered-extensions (string list of extensions that will not be exported
To avoid incoherence, the rule can be that :
if both iidm.import.xml.extensions and iidm.import.xml.filtered-extensions are used,
then ignore the iidm.import.xml.filtered-extensions has priority hence the extension will not be imported
Same behavior for export filter list.
Signed-off-by: Fabrice Buscaylet <fabrice.buscaylet@artelys.com>

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractTreeDataImporter.java:
```diff
@@ -63,6 +65,10 @@ public abstract class AbstractTreeDataImporter implements Importer {
             EXTENSIONS_SUPPLIER.get().getProviders().stream().map(ExtensionProvider::getExtensionName).collect(Collectors.toList()));
 
+    private static final Parameter EXTENSIONS_FILTERED_LIST_PARAMETER
+            = new Parameter(EXTENSIONS_FILTERED_LIST, ParameterType.STRING_LIST, "The list of extension files that will be ignored and not imported ", null,
```
```suggestion
            = new Parameter(EXTENSIONS_FILTERED_LIST, ParameterType.STRING_LIST, "The list of extensions that will be ignored and not imported", null,
```

snoopfab replied:
do you want me to change also EXTENSIONS_LIST comment ? "The list of extension files " to "The list of extensions" ?

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractOptions.java:
```diff
@@ -54,6 +67,10 @@ public boolean withExtension(String extensionName) {
         return withAllExtensions() || extensions.contains(extensionName);
     }
 
+    public boolean withFilteredExtension(String extensionName) {
+        return filteredExtension != null && filteredExtension.contains(extensionName);
+    }
```
You can avoid adding this method by changing the logic of `withExtension`:

```java
    public boolean withExtension(String extensionName) {
        return (filteredExtension == null || !filteredExtension.contains(extensionName))
            && (withAllExtensions() || extensions.contains(extensionName));
    }
```
It would simplify your checks in `NetworkSerde`

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/NetworkSerDe.java:
```diff
@@ -860,8 +871,8 @@ private static void readExtensions(Network network, String id, NetworkDeserializ
             String extensionName = extensionSerde != null ? extensionSerde.getExtensionName() : extensionSerializationName;
-            if (!context.isIgnoredEquipment(id)
-                    && (context.getOptions().withExtension(extensionName) || context.getOptions().withExtension(extensionSerializationName))) {
+            if (!context.isIgnoredEquipment(id) && ((context.getOptions().withExtension(extensionName) || context.getOptions().withExtension(extensionSerializationName))) &&
+                    !context.getOptions().withFilteredExtension(extensionName) && !context.getOptions().withFilteredExtension(extensionSerializationName)) {
```
With the new logic, the code might stay the same as it was

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/NetworkSerDe.java:
```diff
@@ -211,7 +211,9 @@ private static void writeExtensions(Network n, NetworkSerializerContext context,
             }
             Collection<? extends Extension<? extends Identifiable<?>>> extensions = identifiable.getExtensions().stream()
-                    .filter(e -> canTheExtensionBeWritten(getExtensionSerializer(context.getOptions(), e, extensionsSupplier), context.getVersion(), context.getOptions()))
+                    .filter(e ->
+                            !isExtensionFiltered(getExtensionSerializer(context.getOptions(), e, extensionsSupplier), context.getOptions()) &&
+                            canTheExtensionBeWritten(getExtensionSerializer(context.getOptions(), e, extensionsSupplier), context.getVersion(), context.getOptions()))
```
same here

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/ExportOptions.java:
```diff
@@ -106,6 +106,16 @@ public ExportOptions addExtension(String extension) {
     }
 
+    @Override
+    public ExportOptions addFilteredExtension(String extensionToBeFiltered) {
```
If this method is the same in ExportOptions and in ImportOptions, it could be implemented directly in AbstractOptions instead

snoopfab commented:
renamed setExtensions (resp. setFilteredExtensions) to setIncludedExtensions (resp. setExcludedExtensions)
renamed addExtension (resp. addFilteredExtension) to addIncludedExtension (resp. addExcludedExtension)
when calling setIncludedExtensions (resp. setExcludedExtensions), set the blacklist (resp. whitelist) as null in addition to setting the provided list (ie. reset the other list)
when calling addIncludedExtension (resp. addExcludedExtension):
if a whitelist (resp. blacklist) is defined, add the extension to the list
if a blacklist (resp. whitelist) is defined, remove the extension from the list if it is already present. If it's not present, then do nothing
if no list is already defined, then initialize the whitelist (resp. blacklist) with the provided extension

I Added the following warning  logs when specific conditions are met (for the user to be aware of the previous settings overriding):

for setters methods (setFilteredExtensions,setIncludedExtensions ) :
- Previously excluded extensions {} will be ignored ! (setIncludedExtensions )
- All Extensions will be disabled (when passing en empty set to setIncludedExtensions to allow disabling all extensions
- Previously included extensions {} will be ignored !

for adder methods (addIncludedExtension,addExcludedExtension) :
- Previously excluded extensions {} will be included"
- Extension {} is not already explicitly included {} ! It will still be added to the exclusion list as it may be included after (this one is when you add an exclusion when a list of included extension was defined but doesn't contain the extension to be filtered , in this case the programmer still has the possibility to add the extension in the included list )

- Adapted and added tests to reflect the new expected behavior
- Adapted the documentation

for ImportOptions / ExportOptions Parameters:
now throw an ConfigurationException when both include and exclude extensions list are defined at the same time

snoopfab commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractOptions.java:
```diff
@@ -18,40 +22,96 @@
+    }
+
+    public T addExcludedExtension(String extension) throws PowsyblException {
+        if (includedExtensions != null && !includedExtensions.contains(extension)) {
+            LOGGER.warn("Extension {} is not already explicitly included {} ! It will still be added to the exclusion list as it may be included after", extension, includedExtensions);
+        }
+        if (includedExtensions != null) {
+            includedExtensions.remove(extension);
+        }
+        if (excludedExtensions != null) {
+            excludedExtensions.add(extension);
+        } else {
+            this.excludedExtensions = Sets.newHashSet(extension);
+        }
```
With the proposed changes in addExcludedExtension if an includedExtension was defined before, the extensions could never be added to excludedExtensions and without any warning after so i propose the following change to cover this issue :
```
if (includedExtensions != null) {
    if (includedExtensions.remove(extension)) {
        LOGGER.warn("Previously included extensions {} will now be excluded", extension);
    }
} 
if (excludedExtensions == null) {
    excludedExtensions = new HashSet<>();
}
excludedExtensions.add(extension);

```

rolnico replied:
If `includedExtensions` was previously defined, it already means that only the extensions from that list will be included. In other word: all the extensions not in this list are excluded.
Therefore, if you add a new excluded extension via `addExcludedExtension` while `includedExtensions` is defined, you just have to remove the extension from `includedExtensions` and it will not be included (and therefore it will be excluded).

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractOptionExtensionChecker.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * This delegate class avoid code duplication in AbstractTreeDataImporter and AbstractTreeDataExporter classes
+ */
+public abstract class AbstractOptionExtensionChecker {
```
Since it's an utilitary class, I would suggest to make it `final` and not `abstract`

rolnico replied:
Just rename the class to something that does not start with `Abstract`

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractOptionExtensionChecker.java:
```diff
@@ -0,0 +1,59 @@
+        }
+        if (includedExtensions != null) {
+            options.setIncludedExtensions(new HashSet<>(includedExtensions));
+            if (shouldWarnOnInclusionEmptiness && includedExtensions.isEmpty()) {
+                LOGGER.info("All extensions are excluded.");
+                return false;
```
if `shouldWarnOnInclusionEmptiness` is false, you still returns true, even if `includedExtensions.isEmpty()` is true.
Should you not do something like this:
```suggestion
            if (includedExtensions.isEmpty()) {
                if (shouldWarnOnInclusionEmptiness) {
                    LOGGER.warn("No extensions are included.");
                }
                return false;
```
Maybe there is a way to have a logic a bit more clean with less ifs

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractTreeDataExporter.java:
```diff
@@ -194,12 +198,14 @@ private ExportOptions createExportOptions(Properties parameters) {
                 .setFormat(getTreeDataFormat())
                 .setWithAutomationSystems(Parameter.readBoolean(getFormat(), parameters, WITH_AUTOMATION_SYSTEMS_PARAMETER, defaultValueConfig));
-        addExtensionsVersions(parameters, options);
+        boolean someExtensionsShouldBeIncluded = AbstractOptionExtensionChecker.getAndCheckExtensionsToInclude(parameters, options, getFormat(), defaultValueConfig, EXTENSIONS_INCLUDED_LIST_PARAMETER, EXTENSIONS_EXCLUDED_LIST_PARAMETER, true);
```
Import directly the static method

```suggestion
        boolean someExtensionsShouldBeIncluded = getAndCheckExtensionsToInclude(parameters, options, getFormat(), defaultValueConfig, EXTENSIONS_INCLUDED_LIST_PARAMETER, EXTENSIONS_EXCLUDED_LIST_PARAMETER, true);
```

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractTreeDataImporter.java:
```diff
@@ -172,12 +179,15 @@ public Network importData(ReadOnlyDataSource dataSource, NetworkFactory networkF
                 .setWithAutomationSystems(Parameter.readBoolean(getFormat(), parameters, WITH_AUTOMATION_SYSTEMS_PARAMETER, defaultValueConfig))
                 .setMissingPermanentLimitPercentage(Parameter.readDouble(getFormat(), parameters, MISSING_PERMANENT_LIMIT_PERCENTAGE_PARAMETER, defaultValueConfig))
                 .setMinimalValidationLevel(Parameter.readString(getFormat(), parameters, MINIMAL_VALIDATION_LEVEL_PARAMETER, defaultValueConfig));
+        AbstractOptionExtensionChecker.getAndCheckExtensionsToInclude(parameters, importOptions, getFormat(), defaultValueConfig, EXTENSIONS_INCLUDED_LIST_PARAMETER, EXTENSIONS_EXCLUDED_LIST_PARAMETER, false);
```
same here

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/AbstractTreeDataExporter.java:
```diff
@@ -174,13 +180,11 @@ private void addExtensionsVersions(Properties parameters, ExportOptions options)
             Parameter parameter = new Parameter("iidm.export.xml." + extensionName + ".version",
                     ParameterType.STRING, "Version of " + extensionName, null);
             String extensionVersion = Parameter.readString(getFormat(), parameters, parameter, defaultValueConfig);
-            if (extensionVersion != null) {
```
Why are you removing the check on `extensionVersion`?

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/NetworkSerDe.java:
```diff
@@ -231,6 +234,19 @@ private static boolean ignoreEquipmentAtExport(Identifiable<?> identifiable, Net
 
+    private static boolean isExtensionExcluded(ExtensionSerDe extensionSerDe, ExportOptions options) {
+        if (extensionSerDe == null) {
+            return false;
```
should not the extension be excluded if the `extensionSerDe` is null?
The line seems to be covered but the tests works both when I return true and false so there might be an issue. You may want to have a dedicated test for this

rolnico commented on iidm/iidm-serde/src/main/java/com/powsybl/iidm/serde/NetworkSerDe.java:
```diff
@@ -231,6 +234,19 @@ private static boolean ignoreEquipmentAtExport(Identifiable<?> identifiable, Net
                 || identifiable instanceof OverloadManagementSystem && !context.getOptions().isWithAutomationSystems();
     }
 
+    private static boolean isExtensionExcluded(ExtensionSerDe extensionSerDe, ExportOptions options) {
+        if (extensionSerDe == null) {
+            return false;
+        }
+        return options.getIncludedExtensions().filter(Set::isEmpty).isPresent() ||
+                options.getExcludedExtensions()
+                        .map(extensions -> extensions.contains(extensionSerDe.getExtensionName()))
+                        .orElse(false) &&
+                        options.getIncludedExtensions()
+                                .map(extensions -> !extensions.contains(extensionSerDe.getExtensionName()))
+                                .orElse(false);
+    }
```
You could use the method `withExtension` here, and inverse the logic:
```suggestion
    private static boolean isExtensionIncluded(ExtensionSerDe extensionSerDe, ExportOptions options) {
        if (extensionSerDe == null) {
            return false;
        }
        return options.withExtension(extensionSerDe.getExtensionName());
    }
```

In that case you have to remove a "!" where this method is used
